// @milkdown/crepe/lib/esm/index-BUBXHhHG.js
import { configureLinkTooltip, linkTooltipConfig, linkTooltipPlugin } from '@milkdown/kit/component/link-tooltip';
import { c as confirmIcon } from './confirm-DtE-HkVd.js';
import { e as editIcon } from './edit-D--t_JAb.js';
import { r as removeIcon } from './remove-D3g0Cd6_.js';

const copyIcon = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    height="24px"
    viewBox="0 -960 960 960"
    width="24px"
    fill="none"
  >
    <path
      d="M360-240q-33 0-56.5-23.5T280-320v-480q0-33 23.5-56.5T360-880h360q33 0 56.5 23.5T800-800v480q0 33-23.5 56.5T720-240H360Zm0-80h360v-480H360v480ZM200-80q-33 0-56.5-23.5T120-160v-560h80v560h440v80H200Zm160-240v-480 480Z"
    />
  </svg>
`;

const defineFeature = (editor, config) => {
  editor.config(configureLinkTooltip).config((ctx) => {
    ctx.update(linkTooltipConfig.key, (prev) => {
      var _a, _b, _c, _d, _e, _f;
      return {
        ...prev,
        linkIcon: (_a = config == null ? void 0 : config.linkIcon) != null ? _a : () => copyIcon,
        editButton: (_b = config == null ? void 0 : config.editButton) != null ? _b : () => editIcon,
        removeButton: (_c = config == null ? void 0 : config.removeButton) != null ? _c : () => removeIcon,
        confirmButton: (_d = config == null ? void 0 : config.confirmButton) != null ? _d : () => confirmIcon,
        inputPlaceholder: (_e = config == null ? void 0 : config.inputPlaceholder) != null ? _e : "Paste link...",
        onCopyLink: (_f = config == null ? void 0 : config.onCopyLink) != null ? _f : () => {
        }
      };
    });
  }).use(linkTooltipPlugin);
};

export { defineFeature };
